### PR TITLE
[date] header_only option

### LINKED ISSUE #1772: [package] date/2.4.1: header_only option
Please add a header_only option to 'date' recipe.

### PR DESCRIPTION
**Pull request by jeremy-coulon**
Specify library name and version:  **date/2.4.1**

closes #1772

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by jeremy-coulon** [comment]
This PR solves issue #1772

File: recipes/date/all/test_package/conanfile.py [hunk 1]
```diff
@@ -8,10 +8,11 @@ class TestPackageConan(ConanFile):
 
     def build(self):
         cmake = CMake(self)
+        cmake.definitions["DATE_HEADER_ONLY"] = self.options["date:header_only"]
```
Comment (jgsogo) [inline recipes/date/all/test_package/conanfile.py:11]: Add this to `cpp_info.defines` in the recipe. Other consumers will need to define this too.

Comment (jeremy-coulon) [reply recipes/date/all/test_package/conanfile.py:11]: No they won't need it. I just invented this define in order to disable tests of some features that doesn't work in header_only mode.

Comment (uilianries) [reply recipes/date/all/test_package/conanfile.py:11]: I've checked DATE_HEADER_ONLY, curl and pthread. They are not listed on headers.

File: recipes/date/all/conanfile.py [hunk 2]
```diff
@@ -63,18 +67,34 @@ def source(self):
     def build(self):
         for patch in self.conan_data["patches"][self.version]:
             tools.patch(**patch)
+        if self.options.header_only:
+            self.output.warn("Header only package, skipping build")
```
Comment (prince-chrismc) [inline recipes/date/all/conanfile.py:71]:
```suggestion
            self.output.info("Header only package, skipping build")
```

File: recipes/date/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -7,3 +7,10 @@ conan_basic_setup()
 add_executable(${PROJECT_NAME} test_package.cpp)
 target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
 set_property(TARGET ${PROJECT_NAME} PROPERTY CXX_STANDARD 11)
+if(DATE_HEADER_ONLY)
+  set_target_properties(
+	  ${PROJECT_NAME}
+	  PROPERTIES
+	  COMPILE_DEFINITIONS "DATE_HEADER_ONLY"
+	  )
```
Comment (madebr) [inline recipes/date/all/test_package/CMakeLists.txt:15]:
```suggestion
  target_compile_definitions(${PROJECT_NAME} PRIVATE DATE_HEADER_ONLY)
```
Using `set_target_compile_definitions` overwrites all earlier compile definitions.
It does not really matter here, but is weird to do here.

See this example:
```
cmake_minimum_required(VERSION 3.17)
project(p)
file(WRITE "${CMAKE_BINARY_DIR}/src.cpp" "
#include <iostream>
int main() {
#ifndef target_compile_definitions
    std::cout << \"target_compile_definitions NOT DEFINED\\n\";
#endif
#ifndef set_target_properties
    std::cout << \"set_target_properties NOT DEFINED\\n\";
#endif
    return 0;
}")
add_executable(main src.cpp)
target_compile_definitions(main PRIVATE target_compile_definitions)
set_target_properties(main PROPERTIES COMPILE_DEFINITIONS set_target_properties)
```

Running the executable will print `target_compile_definitions NOT DEFINED`.

**Pull request merged**